core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/RareTopNOperator.java

```diff
@@ -0,0 +1,218 @@
+import com.amazon.opendistroforelasticsearch.sql.data.model.ExprValue;
+import com.amazon.opendistroforelasticsearch.sql.expression.Expression;
+import com.amazon.opendistroforelasticsearch.sql.storage.bindingtuple.BindingTuple;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Streams;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.ToString;
+
+/**
+ * Group the all the input {@link BindingTuple} by {@link RareTopNOperator#groupByExprList},
+ * Calculate the rare result by using the {@link RareTopNOperator#fieldExprList}.
+ */
+@ToString
+@EqualsAndHashCode
+public class RareTopNOperator extends PhysicalPlan {
+
+  @Getter
+  private final PhysicalPlan input;
+
+  /**
+   * If rareTopFlag is true, result for top will be returned.
+   * If rareTopFlag is false, result for rare will be returned.
+   */
+  @Getter
+  private final Boolean rareTopFlag;
+  @Getter
+  private final Integer noOfResults;
+  @Getter
+  private final List<Expression> fieldExprList;
+  @Getter
+  private final List<Expression> groupByExprList;
+
+  @EqualsAndHashCode.Exclude
+  private final Group group;
+  @EqualsAndHashCode.Exclude
+  private Iterator<ExprValue> iterator;
+
+  private static final Integer DEFAULT_NO_OF_RESULTS = 10;
+
+
+  public RareTopNOperator(PhysicalPlan input, Boolean rareTopFlag, List<Expression> fieldExprList,
+      List<Expression> groupByExprList) {
+    this(input, rareTopFlag, DEFAULT_NO_OF_RESULTS, fieldExprList, groupByExprList);
+  }
+
+  /**
+   * RareTopNOperator Constructor.
+   *
+   * @param input           Input {@link PhysicalPlan}
+   * @param rareTopFlag     Flag for Rare/TopN command.
+   * @param noOfResults     Number of results
+   * @param fieldExprList   List of {@link Expression}
+   * @param groupByExprList List of group by {@link Expression}
+   */
+  public RareTopNOperator(PhysicalPlan input, Boolean rareTopFlag, int noOfResults,
+      List<Expression> fieldExprList,
+      List<Expression> groupByExprList) {
+    this.input = input;
+    this.rareTopFlag = rareTopFlag;
+    this.noOfResults = noOfResults;
+    this.fieldExprList = fieldExprList;
+    this.groupByExprList = groupByExprList;
+    this.group = new Group();
+  }
+
+  @Override
+  public <R, C> R accept(PhysicalPlanNodeVisitor<R, C> visitor, C context) {
+    return visitor.visitRareTopN(this, context);
+  }
+
+  @Override
+  public List<PhysicalPlan> getChild() {
+    return Collections.singletonList(input);
+  }
+
+  @Override
+  public boolean hasNext() {
+    return iterator.hasNext();
```

Comment:
Let's add a precondition check here and on next() to verify open() has been called.

Reply:
Can you elaborate more? Like what should we return if a condition fails? Maybe Null? 
I have simply followed the logic of AggregationOperator here. It might not be required since open() is updating an iterator and if open() isn't called then hasNext() will simply return false since there are no elements. And when hasNext() will return false, next() won't be called.

Reply:
I would add Precondition.checkNotNull(iterator) here. But really depends on what coding style is preferred.